# Homework 4 - pandas (Cas Carroll) 
### due 3/1/2022   

Use this notebook and prompts to complete the homework. Throughout there will be hints and some code provided  

### Things you will need:  
- Install os, NumPy, pandas  
- states_covid.csv  
- Bloom_etal_2018_Reduced_Dataset.csv  
- logfiles.tgz (or some other multiple file dataset)  

*NOTE*: Make sure your PATH is correct  

**import packages & check required datasets**   

In [5]:
import os
import numpy as np
import pandas as pd

In [6]:
PATH = '/Users/cascarroll/Desktop/spring22/BIOL792_DataScience/Data_Science_For_Biology_II/Part.3.PythonProgramming/Pandas' #CHANGE

In [7]:
assert os.path.exists(os.path.join(PATH,'states_covid.csv')), 'states_covid.csv does not exist' 

In [8]:
assert os.path.exists(os.path.join(PATH,'Bloom_etal_2018_Reduced_Dataset.csv')), 'Bloom_etal_2018_Reduced_Dataset.csv does not exist'

## Task 1 - DataFrame manipulation  

Using **states_covid.csv**, we are going to read the data in as a DataFrame to manipulate, subset, and filter in various ways.   

### Task 1.1 

Read in states_covid.csv with date as a "date" dtype, and only columns consisting of the hospitalization (4 col), ICU (2 col), and Ventilators (2 col)  

In [9]:
states_covid_df = pd.read_csv('states_covid.csv')
# reading in covid csv

In [10]:
states_covid_df.columns
# looking at all the column names

Index(['date', 'state', 'death', 'deathConfirmed', 'deathIncrease',
       'deathProbable', 'hospitalized', 'hospitalizedCumulative',
       'hospitalizedCurrently', 'hospitalizedIncrease', 'inIcuCumulative',
       'inIcuCurrently', 'negative', 'negativeIncrease',
       'negativeTestsAntibody', 'negativeTestsPeopleAntibody',
       'negativeTestsViral', 'onVentilatorCumulative', 'onVentilatorCurrently',
       'positive', 'positiveCasesViral', 'positiveIncrease', 'positiveScore',
       'positiveTestsAntibody', 'positiveTestsAntigen',
       'positiveTestsPeopleAntibody', 'positiveTestsPeopleAntigen',
       'positiveTestsViral', 'recovered', 'totalTestEncountersViral',
       'totalTestEncountersViralIncrease', 'totalTestResults',
       'totalTestResultsIncrease', 'totalTestsAntibody', 'totalTestsAntigen',
       'totalTestsPeopleAntibody', 'totalTestsPeopleAntigen',
       'totalTestsPeopleViral', 'totalTestsPeopleViralIncrease',
       'totalTestsViral', 'totalTestsViralIncrease'

In [123]:
states_covid_sub_df = pd.read_csv('states_covid.csv',usecols=['date', 'state','hospitalized', 'hospitalizedCumulative', 'hospitalizedCurrently', 'hospitalizedIncrease', 'inIcuCumulative', 'inIcuCurrently', 'onVentilatorCumulative', 'onVentilatorCurrently'],parse_dates=['date'],infer_datetime_format=True)
# reading in the csv with only the columns I want (essentially subsetting)
states_covid_sub_df.head()
# printing the first few lines to make sure the first line worked

date state  hospitalized  hospitalizedCumulative  \
0 2021-02-23    AK        1260.0                  1260.0   
1 2021-02-23    AL       45250.0                 45250.0   
2 2021-02-23    AR       14617.0                 14617.0   
3 2021-02-23    AS           NaN                     NaN   
4 2021-02-23    AZ       57072.0                 57072.0   

   hospitalizedCurrently  hospitalizedIncrease  inIcuCumulative  \
0                   38.0                     9              NaN   
1                  762.0                   122           2632.0   
2                  545.0                    47              NaN   
3                    NaN                     0              NaN   
4                 1515.0                    78              NaN   

   inIcuCurrently  onVentilatorCumulative  onVentilatorCurrently  
0             NaN                     NaN                    5.0  
1             NaN                  1497.0                    NaN  
2           204.0                  1505.0                   99.0  
3             NaN                     NaN                    NaN  
4           447.0                     NaN                  266.0

### Task 1.2 

For each of the following catergories: *currently* hospitalized, *currently* in the ICU, and *currently* on ventilation...  
Find the 5 states with the greatest numbers in each catergory and list them in order.      
*hint*: sort_values, unique  

In [12]:
# Below two lines accomplished the task, but I wanted to try some other methods
states_covid_sub_df.sort_values(['hospitalizedCurrently','state'], ascending=False, inplace=True)
# sorted values by hospitalizedCurrently and state
states_covid_sub_df['state'].unique()
# printed unique values for state (in the same order as the descending data for hospitalizedCurrently)


# Below is a bunch of different things I tried - with the loops, I am trying to make a list of the states from the
#sorted dataset...

#stateRankHos=[]
#for column in states_covid_sub_df[['state', 'hospitalizedCurrently']]:
#    state=states_covid_sub_df['state']
#    hospitalized=states_covid_sub_df['hospitalizedCurrently']
#    if state =='CA':
#        stateRankHos.append(state.values, hospitalized.values)
#    elif state != 'CA':
#        stateRankHos.append(state.values, hospitalized.values)

#for column in states_covid_sub_df[['state', 'hospitalizedCurrently']]:
#    state=states_covid_sub_df['state']
#    hospitalized=states_covid_sub_df['hospitalizedCurrently']
#    values= states_covid_sub_df[column]
#    print('Column:', column)
#    print(states_covid_sub_df['state'].unique('Contents:', values.values))
#    print(states_covid_sub_df['state'].unique)


# Below is the way I am most proud of!
HospitalizedMax=states_covid_sub_df.groupby(['state']).max(['hospitalizedCurrently'])
# grouping by state and only taking the max value for hospitalizedCurrently for each state
HospitalizedMax.sort_values(['hospitalizedCurrently'], ascending=False, inplace=True)
# sorting the values so that the largest number is in the top row and smallest is in the bottom row
#HospitalizedMax
# prints the whole list we made
HospitalizedMax.head()
# prints only the info we want (top 5)

######## Code is the same for the next two tasks with column names changed #########

# Top 5 states: CA, NY, TX, FL, NJ

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
CA              NaN                     NaN                22851.0   
NY          89995.0                 89995.0                18825.0   
TX              NaN                     NaN                14218.0   
FL          79410.0                 79410.0                 9520.0   
NJ          63190.0                 63190.0                 8270.0   

       hospitalizedIncrease  inIcuCumulative  inIcuCurrently  \
state                                                          
CA                        0              NaN          4971.0   
NY                     3424              NaN          5225.0   
TX                        0              NaN          3686.0   
FL                      623              NaN             NaN   
NJ                    16373              NaN          2051.0   

       onVentilatorCumulative  onVentilatorCurrently  
state                                                 
CA                        NaN                    NaN  
NY                        NaN                 2425.0  
TX                        NaN                    NaN  
FL                        NaN                    NaN  
NJ                        NaN                 1705.0

In [13]:
IcuMax=states_covid_sub_df.groupby(['state']).max(['inIcuCurrently'])
IcuMax.sort_values(['inIcuCurrently'], ascending=False, inplace=True)
#IcuMax
IcuMax.head()
# Top 5: NY, CA, TX, NJ, MI

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
NY          89995.0                 89995.0                18825.0   
CA              NaN                     NaN                22851.0   
TX              NaN                     NaN                14218.0   
NJ          63190.0                 63190.0                 8270.0   
MI              NaN                     NaN                 4326.0   

       hospitalizedIncrease  inIcuCumulative  inIcuCurrently  \
state                                                          
NY                     3424              NaN          5225.0   
CA                        0              NaN          4971.0   
TX                        0              NaN          3686.0   
NJ                    16373              NaN          2051.0   
MI                        0              NaN          1663.0   

       onVentilatorCumulative  onVentilatorCurrently  
state                                                 
NY                        NaN                 2425.0  
CA                        NaN                    NaN  
TX                        NaN                    NaN  
NJ                        NaN                 1705.0  
MI                        NaN                 1441.0

In [20]:
VentilatorMax=states_covid_sub_df.groupby(['state']).max(['onVentilatorCurrently'])
VentilatorMax.sort_values(['onVentilatorCurrently'], ascending=False, inplace=True)
#VentilatorMax
VentilatorMax.head()
# Top 5: NY, NJ, MI, OH, AZ

hospitalized  hospitalizedCumulative  hospitalizedCurrently  \
state                                                                
NY          89995.0                 89995.0                18825.0   
NJ          63190.0                 63190.0                 8270.0   
MI              NaN                     NaN                 4326.0   
OH          49651.0                 49651.0                 5296.0   
AZ          57072.0                 57072.0                 5082.0   

       hospitalizedIncrease  inIcuCumulative  inIcuCurrently  \
state                                                          
NY                     3424              NaN          5225.0   
NJ                    16373              NaN          2051.0   
MI                        0              NaN          1663.0   
OH                      657           7073.0          1311.0   
AZ                     5173              NaN          1183.0   

       onVentilatorCumulative  onVentilatorCurrently  
state                                                 
NY                        NaN                 2425.0  
NJ                        NaN                 1705.0  
MI                        NaN                 1441.0  
OH                        NaN                  863.0  
AZ                        NaN                  821.0

### Task 1.3 

Find the date in which each state crossed 1000 cumulative hospitilized covid patients

In [137]:
threshold = states_covid_sub_df.query('hospitalizedCumulative > 1000.00')
# only gives values over 1000 for hospitalizedCumulative
#threshold.dtypes
# looking at the data types for each column just to check
threshold.groupby(['state'], as_index=False).agg('min')
# grouping the data by state and aggregating by minimum to so we only get the minimum value for each state

# Tried below, but it drops the date column, so it doesn't actually solve the problem...:
#threshold.groupby(['state']).min(['hospitalizedCumulative'])

state       date  hospitalized  hospitalizedCumulative  \
0     AK 2020-12-29        1004.0                  1004.0   
1     AL 2020-05-01        1008.0                  1008.0   
2     AR 2020-06-15        1003.0                  1003.0   
3     AZ 2020-04-11        1037.0                  1037.0   
4     CO 2020-04-07        1079.0                  1079.0   
5     CT 2020-05-01        7758.0                  7758.0   
6     FL 2020-04-02        1123.0                  1123.0   
7     GA 2020-04-02        1056.0                  1056.0   
8     HI 2020-10-18        1001.0                  1001.0   
9     ID 2020-08-11        1006.0                  1006.0   
10    IN 2020-05-08        4389.0                  4389.0   
11    KS 2020-06-17        1011.0                  1011.0   
12    KY 2020-04-18        1008.0                  1008.0   
13    MA 2020-04-04        1370.0                  1370.0   
14    MD 2020-04-06        1059.0                  1059.0   
15    ME 2020-12-24        1015.0                  1015.0   
16    MN 2020-04-30        1044.0                  1044.0   
17    MS 2020-04-26        1001.0                  1001.0   
18    MT 2020-10-17        1001.0                  1001.0   
19    ND 2020-10-08        1016.0                  1016.0   
20    NE 2020-06-12        1029.0                  1029.0   
21    NH 2021-01-25        1007.0                  1007.0   
22    NJ 2020-05-26       16373.0                 16373.0   
23    NM 2020-05-20        1139.0                  1139.0   
24    NY 2020-03-21        1531.0                  1531.0   
25    OH 2020-04-04        1006.0                  1006.0   
26    OK 2020-06-03        1003.0                  1003.0   
27    OR 2020-06-25        1006.0                  1006.0   
28    PA 2020-04-04        1004.0                  1004.0   
29    RI 2020-05-03        1013.0                  1013.0   
30    SC 2020-05-01        1110.0                  1110.0   
31    SD 2020-08-29        1006.0                  1006.0   
32    TN 2020-04-29        1013.0                  1013.0   
33    UT 2020-06-13        1012.0                  1012.0   
34    VA 2020-04-16        1048.0                  1048.0   
35    WA 2020-05-21        3125.0                  3125.0   
36    WI 2020-04-14        1049.0                  1049.0   
37    WY 2020-12-18        1001.0                  1001.0   

    hospitalizedCurrently  hospitalizedIncrease  inIcuCumulative  \
0                    32.0                     0              NaN   
1                   420.0                     0            335.0   
2                   206.0                    -4              NaN   
3                   468.0                 -2316              NaN   
4                   211.0                   -13              NaN   
5                    42.0                 -4124              NaN   
6                  2005.0                     0              NaN   
7                   783.0                     0           1057.0   
8                    37.0                   -14              NaN   
9                   105.0                    -1            282.0   
10                  595.0                   -13            990.0   
11                  140.0                     0            333.0   
12                  263.0                    -5            514.0   
13                  290.0                   -91              NaN   
14                  281.0                     0              NaN   
15                   67.0                    -2              NaN   
16                  227.0                     0            358.0   
17                  465.0                     0              NaN   
18                   78.0                    -8              NaN   
19                   31.0                    -5            238.0   
20                   95.0                   -48              NaN   
21                  109.0                    -1            324.0   
22                  349.0                    -6            

## Task 2 - DataFrame summarizing  

Using **Bloom_etal_2018_Reduced_Dataset.csv**, we are going to do more dataframe manipulation and subsetting and summarizing    

### Task 2.1 

Read in Bloom_etal_2018_Reduced_Dataset.csv and create two new columns ('genus','species') that consists of the column *taxa* split at the underscore. Print out the head of this new dataframe and the number of unique genera**     

*hint:* pd.str.split(,expand=True)  

for example:  

| taxa | genus | species   
| :------ | :-- | :---   
| Alosa_alabamae | Alosa | alabamae  

In [134]:
df=pd.read_csv('Bloom_etal_2018_Reduced_Dataset.csv',usecols=['taxa'])
# reading in the Bloom csv but only the taxa column
Bloom_split_df=df['taxa'].str.split('_', expand=True)
# splitting the taxa column by the underscore character
Bloom_split_df.columns=(['genus','species','ssp']) # third column is purely for the last entry... not sure what the cf means in the species column and neither did google
# adding column names - see note next to above line for why there are three columns
Bloom_split_df.head()
# making sure the code did what I wanted - it works

genus         species   ssp
0  Alosa        alabamae  None
1  Alosa           alosa  None
2  Alosa          fallax  None
3  Alosa       mediocris  None
4  Alosa  pseudoharengus  None

In [45]:
len(Bloom_split_df['genus'].unique()) #done 
# getting the 'length' of the list of unique genera - basically just counts the number of unique values in the 'genus' column

34

### Task 2.2 

Create a new dataframe with the mean *logbodysize* and *trophicposition* of each genera. Sort this data frame by the largest body size. Print the head of this dataframe.

In [99]:
fishStats_df=pd.read_csv('Bloom_etal_2018_Reduced_Dataset.csv', usecols=['taxa', 'logbodysize', 'trophic_position'])
# reading in the csv with only the columns we want
fishStats_df[['genus','species','subspecies']] = fishStats_df['taxa'].str.split("_",expand=True)
# splitting the taxa column by the underscore character into three columns
fishStats_subdf=fishStats_df[['logbodysize','trophic_position','genus']]
# subsetting the data again - only taking the body size, trophic position, and genus columns
fishMeans_df = fishStats_subdf.groupby('genus',as_index=True).agg(['mean','count'])
# grouping by genus, and getting the mean values for trophic position/body size and count values for the number of fish in each genera in this dataset
fishMeans_df.columns = fishMeans_df.columns.map('_'.join)
# cleaning up our data frame - so there are not column names with sub column names
fishMeans_df.sort_values('logbodysize_mean', ascending=False,inplace=True)
# sorting values so the largest body size is in the top row, smallest is in the bottom row
fishMeans_df
# prints the data frame - answers to the question are below the data frame

logbodysize_mean  logbodysize_count  trophic_position_mean  \
genus                                                                        
Tenualosa               1.778151                  1               0.462398   
Alosa                   1.739062                  6               0.540815   
Coilia                  1.544068                  1               0.477121   
Ethmalosa               1.544068                  1               0.397940   
Potamalosa              1.505150                  1               0.518514   
Hilsa                   1.387390                  1               0.462398   
Anodontostoma           1.342423                  1               0.447158   
Engraulis               1.190332                  1               0.462398   
Cetengraulis            1.113943                  1               0.322219   
Pellona                 1.113943                  1               0.623249   
Anchovia                1.093422                  1               0.477121   
Thryssa                 1.041393                  1               0.447158   
Pellonula               1.004300                  2               0.504938   
Escualosa               1.000000                  1               0.505150   
Gilchristella           1.000000                  1               0.491362   
Stolothrissa            1.000000                  1               0.431364   
Hyperlophus             1.000000                  1               0.531479   
Stolephorus             0.973128                  1               0.518514   
Chirocentrodon          0.954243                  1               0.518514   
Encrasicholina          0.903090                  1               0.518514   
Clupeoides              0.903090                  1               0.491362   
Ramnogaster             0.903090                  1               0.531479   
Rhinosardinia           0.903090                  1               0.531479   
Spratelloides           0.845098                  1               0.491362   
Anchoa                  0.843681                  7               0.522218   
Jenkinsia               0.838849                  1               0.518514   
Microthrissa            0.778151                  1               0.491362   
Anchoviella             0.775207                  5               0.496877   
Ehirava                 0.698970                  1               0.491362   
Sauvagella              0.672098                  1               0.491362   
Clupeichthys            0.602060                  1               0.477121   
Sierrathrissa           0.477121                  1               0.491362   
Sundasalanx             0.447158                  1               0.491362   
Amazonsprattus          0.301030                  1               0.531479   

                trophic_position_count  
genus                                   
Tenualosa                            1  
Alosa                                6  
Coilia                               1  
Ethmalosa                            1  
Potamalosa                           1  
Hilsa                                1  
Anodontostoma                        1  
Engraulis                            1  
Cetengraulis                         1  
Pellona                              1  
Anchovia                             1  
Thryssa                              1  
Pellonula                            2  
Escualosa                            1  
Gilchristella                        1  
Stolothrissa                         1  
Hyperlophus                          1  
Stolephorus                          1  
Chirocentrodon                       1  
Encrasicholina                       1  
Clupeoides                           1  
Ramnogaster                          1  
Rhinosardinia                        1  
Spratelloides                        1  
Anchoa                               7  
Jenkinsia                            1  
Microthrissa                         1  
Anchoviella       

Which genera is the smallest and largest?

Largest = Tenualosa,
Smallest = Amazonsprattus

What is the trophic position of the smallest and largest?  

Trophic position of the largest = 0.462398,
Trophic position of the smallest = 0.531479

## Task 3 - Read in muliple files to a dictionary and make a DataFrame - **OPTIONAL/BONUS**  

### This is not something you are expected to do in this course, but just here to give you an idea of the things that you COULD do. Answers will be posted after due date.  


Using **logfiles**: we are going to do read in each file, get some data, append it to a dictionary to later make into a dataframe.     

**note:** *make sure to unzip logfiles*

In [20]:
cd $PATH

/Users/trevorfaske/Desktop/Classes/DataScience/pandas/Spring22


In [21]:
!tar -xzf logfiles.tgz #unzip logfiles

In [22]:
log_dir = os.path.join(PATH,'logfiles')
assert os.path.exists(log_dir), 'log_dir does not exist'

First step is to find the necessary files. The number of files in the log files is 36, make sure you have that many as well  

In [23]:
!ls -l logfiles/*txt | wc -l 

      36


In [24]:
logfiles = !find $log_dir -name '*txt' #unix command to find files in log_dir directory
logfiles = [os.path.abspath(x) for x in logfiles] #this finds the full path to the file
print(logfiles[1])

/Users/trevorfaske/g/Classes/DataScience/pandas/Spring22/logfiles/1901302121_H8_S_14.txt.txt


In [25]:
assert(len(logfiles)==36), 'Do not have correct number of logfiles'

### Getting a little tricky here  

Read in each of the logfiles, for each file extract:  
- minimum temperature  
- maximum temperature  
- date of minimum temp   
- date of maximum temp   
- mean temp for each file.   

This data should all be appended for a dictionary within a for loop:    
Key should be the file name without the path or .txt extension  
Values should be (minTemp,maxTemp,minDate,maxDate,meanTemp)

I recommend making this work for one file first, then putting the rest in a for loop to do the rest.  

Below is an example of how to read in one file

*hint:* do not read date in as date object

In [41]:
### set up data frame as you read in each file
infile = pd.read_csv(logfiles[0],sep='\t',engine='python')
infile.columns = ['Index','Date','Time','Temp','Type']
infile = infile[['Date','Time','Temp']]
infile.head()

Date         Time  Temp
0  9/16/2013   8:00:00 AM  47.9
1  9/16/2013   8:35:00 AM  48.2
2  9/16/2013   9:10:00 AM  48.7
3  9/16/2013   9:45:00 AM  49.4
4  9/16/2013  10:20:00 AM  50.3

Do everything in steps, make sure it works. Calculate summaries with this one infile:

In [26]:
minTemp =   
maxTemp =  
minDate = infile['Date'][infile['Temp'] == infile['Temp'].min()].unique()[0] #use this for minDate 
maxDate = infile['Date'][infile['Temp'] == infile['Temp'].max()].unique()[0] #use this for maxDate 
meanTemp = 

SyntaxError: invalid syntax (<ipython-input-26-d888355c2242>, line 1)

To get you started, I suggest writing some dummy code in plain words to help outline your for loop:  

In [ ]:
logfiles_dict = {}  
for f in logfiles:  
    #do something   
    #do not read date in as date object  
    #do more something   
    #do other stuff  
    #make print statements EVERYWHERE  
    #append to dict  
    #blahbahblah

Then do the real code below here. You don't need to turnin your thoughts. Just put it in there as a help reminder. Most people all still do this, no matter how advanced they are

**Once you have created a DataFrame with all the logfiles, print the head and save it to an outfile using pd.to_csv() as logfiles_df.csv** 

#### Here is an example of the final product

In [32]:
logfiles_df.head()

minTemp  maxTemp     minDate    maxDate   meanTemp
1901302235_H6_S_14      21.3     65.8   12/6/2013  7/13/2014  35.726191
1901302121_H8_S_14      17.9     73.4  10/28/2013  7/13/2014  34.698866
1901302217_H18_D_14     21.0     82.4  10/28/2013  7/26/2014  33.334564
1901302110_H13_S_14     22.8     60.9   10/4/2013  7/13/2014  32.240032
1901302146_H14_D_14     24.6     66.1   1/21/2014  7/13/2014  34.242075